staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -655,3 +962,127 @@ func (c claims) hasClaim(name string) bool {
 	}
 	return true
 }
+
+type claimMapper struct {
+	compiler authenticationcel.Compiler
+}
+
+func NewClaimMapper(env *celenvironment.EnvSet) *claimMapper {
+	return &claimMapper{
+		compiler: authenticationcel.NewCompiler(env),
+	}
+}
+
+type evaluationActivation struct {
+	claims, userInfo interface{}
+}
+
+// ResolveName implements the Activation interface.
+func (a *evaluationActivation) ResolveName(name string) (interface{}, bool) {
+	switch name {
+	case authenticationcel.ClaimsVarName:
+		return a.claims, true
+	case authenticationcel.UserInfoVarName:
+		return a.userInfo, true
+	default:
+		return nil, false
+	}
+}
+
+// Parent returns the parent of the current activation, may be nil.
+// If non-nil, the parent will be searched during resolve calls.
+func (a *evaluationActivation) Parent() interpreter.Activation {
+	return nil
+}
+
+func (c *claimMapper) compile(expressionAccessors []authenticationcel.ExpressionAccessor, mode celenvironment.Type) (*mapper, error) {
+	compilationResults := make([]authenticationcel.CompilationResult, len(expressionAccessors))
+	for i, expressionAccessor := range expressionAccessors {
+		if expressionAccessor == nil {
+			continue
+		}
+
+		compResult := c.compiler.CompileCELExpression(expressionAccessor, mode)
+		if compResult.Error != nil {
+			return nil, fmt.Errorf("error compiling expression: %v", compResult.Error)
+		}
+
+		compilationResults[i] = c.compiler.CompileCELExpression(expressionAccessor, mode)
+	}
+	return &mapper{
+		compilationResults: compilationResults,
+	}, nil
+}
+
+type mapper struct {
+	compilationResults []authenticationcel.CompilationResult
+}
+
+func (m *mapper) eval(ctx context.Context, c, userInfo interface{}) ([]authenticationcel.EvaluationResult, error) {
+	evaluations := make([]authenticationcel.EvaluationResult, len(m.compilationResults))
+
+	userInfoVal, err := convertObjectToUnstructured(&userInfo)
+	if err != nil {
+		return nil, fmt.Errorf("error converting user info to unstructured: %v", err)
+	}
+
+	va := &evaluationActivation{
+		claims:   c,
+		userInfo: userInfoVal.Object,
+	}
+
+	for i, compilationResult := range m.compilationResults {
+		var evaluation = &evaluations[i]
+		if compilationResult.ExpressionAccessor == nil {
+			continue
+		}
+		evaluation.ExpressionAccessor = compilationResult.ExpressionAccessor
+		if compilationResult.Error != nil {
+			evaluation.Error = &cel.Error{
+				Type:   cel.ErrorTypeInvalid,
+				Detail: fmt.Sprintf("compilation error: %v", compilationResult.Error),
+			}
+			continue
+		}
+		if compilationResult.Program == nil {
+			evaluation.Error = &cel.Error{
+				Type:   cel.ErrorTypeInternal,
+				Detail: "unexpected internal error compiling expression",
+			}
+			continue
+		}
+		t1 := time.Now()
+		evalResult, evalDetails, err := compilationResult.Program.ContextEval(ctx, va)
```

Comment:
Would you consider to leverage the cost system and set the cost limit to prevent too expensive cel expression from running? ref: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#resource-use-by-validation-functions

Reply:
This is provided by the kube-apiserver admin, not a REST API, so I don't think we need to be especially guarded on cost. Also, the cardinality on the input is unknown at compile time, since arbitrary JWT token claim structures can be encountered at runtime, so errors would be at runtime only, not at start time, which is ... bad.